megamek/src/megamek/client/ratgenerator/FactionRecord.java

```diff
@@ -602,72 +603,74 @@ public void writeToXml(PrintWriter pw) {
-                    .collect(Collectors.joining(",")) + "</pctClan>");
+                    .collect(Collectors.joining(",")) + "</pctClan>\n");
         }
         if (pctTech.containsKey(TechCategory.IS_ADVANCED_AERO)
                 && pctTech.get(TechCategory.IS_ADVANCED_AERO).containsKey(era)
                 && (pctTech.get(TechCategory.IS_ADVANCED_AERO).get(era).size() > 0)) {
-            pw.println("\t\t<pctSL unitType='Aero'>"
+            factionRecordBuilder.append("\t\t<pctSL unitType='Aero'>"
                     + pctTech.get(TechCategory.IS_ADVANCED_AERO).get(era).stream().map(Object::toString)
-                    .collect(Collectors.joining(",")) + "</pctSL>");
+                    .collect(Collectors.joining(",")) + "</pctSL>\n");
         }
         if (pctTech.containsKey(TechCategory.CLAN_VEE)
                 && pctTech.get(TechCategory.CLAN_VEE).containsKey(era)
                 && (pctTech.get(TechCategory.CLAN_VEE).get(era).size() > 0)) {
-            pw.println("\t\t<pctClan unitType='Vehicle'>"
+            factionRecordBuilder.append("\t\t<pctClan unitType='Vehicle'>"
                     + pctTech.get(TechCategory.CLAN_VEE).get(era).stream().map(Object::toString)
-                    .collect(Collectors.joining(",")) + "</pctClan>");
+                    .collect(Collectors.joining(",")) + "</pctClan>\n");
         }
         if (pctTech.containsKey(TechCategory.IS_ADVANCED_VEE)
                 && pctTech.get(TechCategory.IS_ADVANCED_VEE).containsKey(era)
                 && (pctTech.get(TechCategory.IS_ADVANCED_VEE).get(era).size() > 0)) {
-            pw.println("\t\t<pctSL unitType='Vehicle'>"
+            factionRecordBuilder.append("\t\t<pctSL unitType='Vehicle'>"
                     + pctTech.get(TechCategory.IS_ADVANCED_VEE).get(era).stream().map(Object::toString)
-                    .collect(Collectors.joining(",")) + "</pctSL>");
+                    .collect(Collectors.joining(",")) + "</pctSL>\n");
         }
         if (era > 3067) {
             if (!isClan() && !key.equals("CGB.FRR") && !key.equals("RA.OA")) {
-                pw.println("\t\t<omniMargin>" + ((era - 3067) / 5) + "</omniMargin>");
-                pw.println("\t\t<techMargin>" + ((era - 3067) / 5) + "</techMargin>");
+                factionRecordBuilder.append("\t\t<omniMargin>" + ((era - 3067) / 5) + "</omniMargin>\n");
+                factionRecordBuilder.append("\t\t<techMargin>" + ((era - 3067) / 5) + "</techMargin>\n");
                 if (era > 3085) {
-                    pw.println("\t\t<upgradeMargin>" + ((era - 3085) / 5) + "</upgradeMargin>");
+                    factionRecordBuilder.append("\t\t<upgradeMargin>" + ((era - 3085) / 5) + "</upgradeMargin>\n");
                 }
             } else {
-                pw.println("\t\t<techMargin>" + ((era - 3067) / 5) + "</techMargin>");
+                factionRecordBuilder.append("\t\t<techMargin>" + ((era - 3067) / 5) + "</techMargin>\n");
             }
         }
```

Comment:
The main reason is on line 702, because sometimes the faction has nothing in its interior and this was producing invalid XML.

Reply:
It was invalid because the attributes were not escaped and several of the names have apostrophes. But self-closing tags are not required for elements without children.`<faction></faction>` and `<faction/>` are equivalent and both perfectly legal. It doesn't really matter one way or te other, but if you're going to use a StringBuilder you should append each string individually rather than using string concatenation in the argument to append. That has the JVM instatiating an unnecessary additional StringBuilder. This also adds an additional newline between the last child and the end tag. You might consider StringJoiner instead.

Reply:
For whatever reason, which I don't feel like investigating at the moment, <faction></faction> throws an "expected element" exception when loading the data. I got rid of the string concatenation where it was relevant.